File: utilities/mico/openbus/common/ClientInterceptor.cpp
/*
** common/ClientInterceptor.cpp
*/

#include "ClientInterceptor.h"

#ifdef VERBOSE
  #include <iostream>
#endif

#include <mico/pi_impl.h>
#include <mico/codec_impl.h>

#include "../../stubs/access_control_service.h"

using namespace openbusidl::acs;
#ifdef VERBOSE
  using namespace std ;
#endif

namespace openbus {
  namespace common {
    ClientInterceptor::ClientInterceptor( CredentialManager* pcredentialManager ) {
    #ifdef VERBOSE
      cout << "\n\n[ClientInterceptor::ClientInterceptor() BEGIN]" << endl ;
    #endif
      credentialManager = pcredentialManager ;
    #ifdef VERBOSE
      cout << "\n\n[ClientInterceptor::ClientInterceptor() END]" << endl ;
    #endif
    }

    ClientInterceptor::~ClientInterceptor() {
    #ifdef VERBOSE
      cout << "\n\n[ClientInterceptor::~ClientInterceptor() BEGIN]" << endl ;
    #endif
    #ifdef VERBOSE
      cout << "\n\n[ClientInterceptor::~ClientInterceptor() END]" << endl ;
    #endif
    }

    void ClientInterceptor::send_request( ClientRequestInfo_ptr ri )
    {
    #ifdef VERBOSE
      cout << "\n\n[ClientInterceptor::send_request() BEGIN]" << endl ;
    #endif
      Credential* c = credentialManager->getValue() ;
      if ( c != NULL ) {
        IOP::ServiceContext sc ;
        sc.context_id = 1234 ;

        MICO::CDREncoder encoder ;

        CORBA::Any any ;
        any <<= *c ;
        CORBA::OctetSeq_var en ;
        PICodec::Codec_impl* codec = new PICodec::Codec_impl ;
        en = codec->encode_value( any ) ;

        sc.context_data = en ;

      #ifdef VERBOSE
        CORBA::ULong z ;
        cout << "[Context Data:]" ;
        for ( z = 0; z < sc.context_data.length(); z++ ) {
          printf( "%u ", sc.context_data[ z ] ) ;
        }
      #endif

        ri->add_request_service_context( sc, true ) ;
      }
    #ifdef VERBOSE
      cout << "\n[ClientInterceptor::send_request() END]" << endl ;
    #endif
    }

    char* ClientInterceptor::name() { return "" ; }
    void ClientInterceptor::destroy() {}
    void ClientInterceptor::send_poll( ClientRequestInfo_ptr ri ) {}
    void ClientInterceptor::receive_reply( ClientRequestInfo_ptr ri ) {}
    void ClientInterceptor::receive_exception( ClientRequestInfo_ptr ri ) {}
    void ClientInterceptor::receive_other( ClientRequestInfo_ptr ri ) {}
  }
}

File: utilities/c#/src/Interceptors/ClientInterceptor.cs
using System;
using System.Collections.Generic;
using System.Text;
using openbusidl.acs;
using omg.org.IOP;
using omg.org.CORBA;
using OpenbusAPI.Logger;

namespace OpenbusAPI.Interceptors
{
  /// <summary>
  /// Representa o interceptador cliente.
  /// Implementa PortableInterceptor.ClientRequestInterceptor.
  /// </summary>
  class ClientInterceptor : InterceptorImpl, omg.org.PortableInterceptor.ClientRequestInterceptor
  {

    #region Fields

    /// <summary>
    /// Instância do barramento.
    /// </summary>
    private Openbus bus;

    #endregion

    #region Contructor

    /// <summary>
    /// Inicializa uma nova instância de OpenbusAPI.Interceptors.ClientInterceptor
    /// </summary>
    /// <param name="codec">Codificador</param>
    public ClientInterceptor(Codec codec)
      : base("ClientInterceptor", codec) {
      this.bus = Openbus.GetInstance();
    }

    #endregion

    #region ClientRequestInterceptor Members

    /// <summary>
    /// Intercepta o request para inserção de informação de contexto.
    /// </summary>
    /// <remarks>Informação do cliente</remarks>
    public void send_request(omg.org.PortableInterceptor.ClientRequestInfo ri) {
      Log.INTERCEPTORS.Debug("executando método: " + ri.operation);
      
      /* Verifica se existe uma credencial para envio */
      Credential credential = bus.Credential;
      if ((credential.identifier == null) || (credential.identifier.Equals(""))) {
        Log.INTERCEPTORS.Info("Sem Credencial!");
        return;
      }

      Log.INTERCEPTORS.Debug("Tem Credencial");

      byte[] value = null;
      try {
        value = this.Codec.encode_value(credential);
      }
      catch {
        Log.INTERCEPTORS.Fatal("Erro na codificação da credencial.");
      }

      ri.add_request_service_context(new ServiceContext(CONTEXT_ID, value), false);
    }

    #endregion

    #region ClientRequestInterceptor Not Implemented

    public void receive_exception(omg.org.PortableInterceptor.ClientRequestInfo ri) {
      //Nada a ser feito;
    }

    public void receive_other(omg.org.PortableInterceptor.ClientRequestInfo ri) {
      //Nada a ser feito;
    }

    public void receive_reply(omg.org.PortableInterceptor.ClientRequestInfo ri) {
      //Nada a ser feito;
    }

    public void send_poll(omg.org.PortableInterceptor.ClientRequestInfo ri) {
      //Nada a ser feito;
    }

    #endregion

  }
}
